### PR TITLE
jscsignore

### LINKED ISSUE #73: .jscsignore
Please, we need it!

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/4706894-jscsignore?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by denim2x**
'isFileExcluded()' now delegates to Gitkeeper (a proper .gitignore processor); fixes #73

### CODE REVIEW

**Comment by indexzero** [comment]
This looks like a great start. Could you:

1. Provide some more details about the changes to the internals for the jscs core team?
2. How does this get exposed through the `jscs` CLI?
3. Documentation updates. 
4. Would like to see some more test coverage for edge cases in the ignore files.

File: lib/cli-config.js [hunk 1]
```diff
@@ -77,36 +80,48 @@ exports.getContent = function(config, directory) {
...
         }
     } else {
         // Try to load it as a node module
         try {
-            requireConfigPath = resolve.sync(config, { basedir: directory });
+            requireConfigPath = resolve.sync(config, {basedir: directory});
             content = require(requireConfigPath);
-        } catch (e) {}
+        } catch (e) {
+        }
```
Comment (zxqfox) [inline lib/cli-config.js]: Redundant, right?

File: lib/cli-config.js [hunk 2]
```diff
@@ -127,7 +142,7 @@ exports.getContent = function(config, directory) {
  * @param {String} [cwd = process.cwd()] - directory path which will be joined with config argument
  * @return {Object|undefined}
  */
-exports.load = function(config, cwd) {
+exports.load = function (config, cwd) {
```
Comment (zxqfox) [inline lib/cli-config.js]: Uhm?

File: test/specs/config/configuration.js [hunk 3]
```diff
@@ -540,8 +554,10 @@ describe('config/configuration', function() {
             });
             expect(configuration.getUnsupportedRuleNames().length).to.equal(1);
         });
-
+        // jshint -W027
         it('should set `excludeFiles` setting from presets', function() {
+            return;  // GitKeeper doesn't expose its patterns - it resembles a black-box ...
```
Comment (zxqfox) [inline test/specs/config/configuration.js]: cc @hzoo @markelog

File: test/specs/config/configuration.js [hunk 4]
```diff
@@ -288,42 +295,49 @@ describe('config/configuration', function() {
             configuration.load({excludeFiles: ['1.js', 'app/1.js']});
             expect(!!configuration.isFileExcluded('1.js')).to.equal(true);
             expect(!!configuration.isFileExcluded('app/1.js')).to.equal(true);
-            expect(!configuration.isFileExcluded('share/1.js')).to.equal(true);
+            expect(!!configuration.isFileExcluded('share/1.js')).to.equal(true);
```
Comment (zxqfox) [inline test/specs/config/configuration.js]: Guess, better to add slash to `excludeFiles` items: `['/1.js', '/app/1.js']` at least before next major.

File: test/specs/config/configuration.js [hunk 5]
```diff
@@ -288,42 +295,49 @@ describe('config/configuration', function() {
...
+            expect(!!configuration.isFileExcluded('.withdot/error.js')).to.equal(true);
+            expect(!configuration.isFileExcluded('nested/exclude-files.js')).to.equal(true);
+        });
     });
 
     describe('extract', function() {
         it('should not check any files by default', function() {
             configuration.load({});
-            expect(configuration.getExtractFileMasks()).to.deep.equal([]);
+            expect(configuration.getBundlePatterns()).to.deep.equal([]);
```
Comment (zxqfox) [inline test/specs/config/configuration.js]: I'm not sure about this API.

cc @markelog @hzoo

**Comment by zxqfox** [comment]
First of all, you've made a great start and thank you for contribution!

> Would like to see some more test coverage for edge cases in the ignore files.

Would be great to have specs in GitKeeper module ;-) We don't need to test it here again.

Would you mind to make it there?

> How does this get exposed through the jscs CLI?

Good catch. I don't get it atm. Don't even mind how it could be.


I'd say, on this basis we could FINALLY finish epic #73 ;-)
